rpc.go

```diff
@@ -126,12 +134,61 @@ func (c *client) SendRPC(rpc hrpc.Call) (msg proto.Message, err error) {
 			serverErrorCount++
 			continue // retry
 		case region.NotServingRegionError:
+			if scan, ok := rpc.(*hrpc.Scan); ok {
+				c.scanRpcScanStats(scan, msg, err, true, start, time.Now())
+			}
 			continue // retry
 		}
+		if scan, ok := rpc.(*hrpc.Scan); ok {
+			c.scanRpcScanStats(scan, msg, err, false, start, time.Now())
+		}
 		return msg, err
 	}
 }
 
+func (c *client) scanRpcScanStats(scan *hrpc.Scan, resp proto.Message, err error,
+	retry bool, start, end time.Time) {
+	if scan.ScanStatsHandler() != nil {
+		stats := &hrpc.ScanStats{}
+		// Update the ScanMetrics if they are being tracked. For ScanStats, these ScanMetrics
+		// are collected per call to SendRPC and therefore may not be reflective of the entire
+		// result of the Scan if the results are split across multiple calls to Scanner.Next().
+		if scan.TrackScanMetrics() {
+			scanres, ok := resp.(*pb.ScanResponse)
+			if !ok {
+				c.logger.Error("got non ScanResponse for ScanRequest")
+			} else {
+				if scanres.ScanMetrics != nil {
+					if stats.ScanMetrics == nil {
+						stats.ScanMetrics = make(map[string]int64)
+					}
+					for _, m := range scanres.ScanMetrics.GetMetrics() {
+						stats.ScanMetrics[m.GetName()] += m.GetValue()
+					}
```

Comment:
This is now being updated in two places, here and in `scanner.fetch`, which I think means the same metrics will be added twice. And as your comment above mentions this is being added up across multiple scans and for the ScanStats we should just have the metrics for one scan.
Perhaps you can just create a local `map[string]int64` for the metrics found for just this one scan and put that in the ScanStats struct. And leave the code in `scanner.fetch` to be responsible for summing the ScanMetrics across multiple scans.

Reply:
For the ScanStats, there is a local map[string]int64 for the ScanStats.ScanMetrics. This is different from the scanner.ScanMetrics that's updated in scanner.fetch. which sums the metrics. I've updated the ScanStats one to not add it like in scanner.fetch.